chapters/edo/main.tex

```diff
@@ -1,2 +1,1193 @@
+increasing the selective pressure (by reducing \(b\)) to mitigate the effect of
+noise in later generations.
+
+In addition to these parameter changes, the fitness function has been altered.
+In this study, every individual is scaled using a min-max scaler so their values
+are in the interval \(\left[0, 1\right]\). This makes the values of the limits
+in~\eqref{eq:uniform} arbitrary and eliminates the pinching effect observed
+in~\cite{Wilde2020:edo} where well-performing individuals were
+disproportionately compact. Following this scaling, the number of
+initialisations for the \(k\)-means algorithm has been increased from ten to 50
+so that there is greater confidence in any given fitness score.
+
+The examples in this study make use of a command-line tool,
+\href{https://github.com/daffidwilde/edolab}{\mintinline{python}{edolab}}, for
+running experiments with the library. This tool allows for a lot of otherwise
+repeated code to be replaced by a simple \emph{experiment} script, configuring
+the parameters of the experiment. Snippet~\ref{snp:inertia} shows the experiment
+script used for this example, and Snippet~\ref{snp:edolab} shows how to use that
+script with the command-line tool. Other than the fitness function definition,
+this script is identical to that of every example in this section.
+
+\begin{listing}
+\begin{sourcepy}
+""" /path/to/experiments/kmeans_inertia.py """
+
+from edo.distributions import Uniform
+from sklearn.cluster import KMeans
+from sklearn.preprocessing import MinMaxScaler
+
+
+def fitness(individual, max_seed=5):
+    """ Return the lowest final inertia of k-means on the individual
+    across the given number of trials with k=3. """
+
+    data = MinMaxScaler().fit_transform(individual.dataframe, copy=True)
+
+    inertias = []
+    for seed in range(max_seeds):
+        km = KMeans(n_clusters=3, random_state=seed).fit(data)
+        inertias.append(km.inertia_)
+
+    return min(inertias)
+
+
+size = 100
+row_limits = [50, 100]
+col_limits = [2, 2]
+max_iter = 100
+best_prop = 0.1
+lucky_prop = 0
+mutation_prob = 0.01
+
+Uniform.param_limits["bounds"] = [0, 1]
+distributions = [Uniform]
+\end{sourcepy}
+\caption{%
+    An abridged version of the experiment configuration script used in the first
+    example
+}\label{snp:inertia}
+\end{listing}
+
+\begin{listing}
+\begin{usagesh}
+> cd /path/to/experiments
+> edolab run --seeds=10 --cores=4 kmeans_inertia.py
+> edolab summarise --tarball kmeans_inertia.py
+\end{usagesh}
+\caption{Example usage of the \mintinline{python}{edolab} command-line tool}
+\label{snp:edolab}
+\end{listing}
+
+Once the EDO algorithm has terminated, a body of datasets, and information about
+those datasets, is recorded. This output is referred to as a \emph{history}.
+Figure~\ref{fig:inertia_progression} shows the progression of the fitness
+function (inertia) and the number of rows at ten generation intervals across the
```

Comment:
Just a thought, not sure if here is the right place but is it worth having a sentence about the pure volume of data that eco can produce? (Not necessarily a good/bad thing but worth being aware of).